### DISCUSSION TITLE
Next Version Updates

### DISCUSSION THREAD

**Discussion by coreybutler** [coreybutler/node-windows#317] [Ideas]
In response to #308, I've reviewed the code pretty extensively. It's... old. I've also reviewed node-mac and node-linux, which are also pretty old code bases. These code bases will continue to work, but I believe there is a better approach.

This library is based on winsw, whereas node-mac directly accesses the OS daemon manager. node-linux only supports systemv (not systemd). The main challenge in maintaining a consistent API is the fact each of these approaches are considerably different... not to mention maintaining 3 repos is getting pretty tired.

At this point, I believe it would be better to abstract the wrapper functionality (i.e. complexity) out of JavaScript and bundle it within the executable itself. This is something that can be done across all operating systems with a Go executable, so I am planning to experiment with that.

Just to pre-answer the question "why rebuild winsw?" (which was asked the last time I brought this up), that project only works on Windows. The goal of this project and its sister projects was always to provide a common API across all operating systems.

If the experiments go well, I intend to deprecate this library (and node-mac/linux) in favor of a new single library focused exclusively on background services/daemons. If these changes happen, it likely won't be for awhile, seeing as I'm still pretty pressed for time.

**Comment by coreybutler** [comment]
Some of the other features of this library (sudo, taskkill, etc) _may_ be extracted into another separate library if there is enough interest. Feel free to voice your interest here by upvoting this comment.

**Reply by coreybutler** [reply]
I've reflected on this since writing the first post. Since I was pinged on Twitter about it today, I decided to post my recent thoughts:

## Standalone Executables == Dead
With the emergence of [Node SEA](https://nodejs.org/api/single-executable-applications.html), there will be a native way to distribute Node.js apps. That is the future of standalone executables, which is mirrored in Deno and Bun.

_However_, SEA executables suffer from some big drawbacks. 

1. The embedded runtime produces large binaries. Last time I tried, a `console.log` app was 90MB.
1. Reliable distro on Windows requires code signing of each binary.

## Decoupling the Process Manager

When I first wrote this module, the intention was to use native process managers provided by the OS. The challenge was registering/unregistering a service (i.e. install/uninstall), which this library does for you. But this library stops at "basic self-introspection". There is much more that could be done, as projects like pm2 have highlighted. However; pm2 has evolved into a full business to support the development efforts required to achieve that.

## Sidecar Strategy w/ New Binary (i.e. my current thinking)
This module uses the sidecar strategy. For the unfamiliar, this strategy employs a single common executable capable of wrapping any app, essentially bootstrapping the startup process. However; the current version basically just packages up winsw, renames it, and provides a config for winsw.

Winsw is limited to Windows. The overarching goal of this project was to provide a consistent management experience on any OS (thus [node-mac](https://github.com/coreybutler/node-mac)/[node-linux](https://github.com/coreybutler/node-linux)). By creating our own dedicated binary (in Go), we could achieve this.

There is a LOT we could do with this. The staples would remain, like exponential backoff, native logging, and of course operating as a native background service. But we could also add capabilities like masked logging (i.e. **** instead of `my_password`), OTel support, RAM/CPU warnings/thresholds, Prometheus support, healthcheck endpoints. I have a long list of other things that could be incorporated into this.

From an administrative perspective, we could resolve a lot of distribution problems for people too. The new company I started, [Author Software](https://linkedin.com/company/authorsoftware), has already resolved much of this via other projects. We code-sign [NVM for Windows](https://github.com/coreybutler/nvm-windows), meaning our certificates are already trusted by millions of people and companies. We could use that to code-sign a sidecar executable to launch Node apps. There are other trust factors under consideration, but this would break through a barrier that is difficult (or sometimes impossible) for many people... especially for those who just want to distribute their apps within their own environment.

Ultimately, I think the concepts above are a better way to resolve the problems node-windows solves.

There is a lot of work that would go into this. The initial stuff I've mentioned here is pretty straightforward, but I can anticipate some of the needs (because I've already encountered them myself). In other words, making this sidecar concept a reality is definitely feasible in a reasonably short period of time. It's everything that comes after it that would be more challenging. I've faced this with several of my OSS projects now. They reach a point where they help, but when enterprises start asking for features, it requires significantly more work. That's why I created Author Software. More likely than not, I would release this as a new project under the Author Software banner, offering the core for free with extended development/capabilities part of a monetized strategy. I'd provide a transition plan for node-windows users and eventually shut this project down.

So, that's where my head is right now. As far as a timeline goes, my focus is on [Runtime](https://github.com/coreybutler/nvm-windows/wiki/Runtime) first. This project would likely come after that.

**Reply by radcapitalist** [reply]
TL;DR: node-windows will go unmaintained for the next three years and then be shut down. Is that a fair characterization?

Eric

**Reply by radcapitalist** [reply]
Another huge drawback of Node SEA is the Common JS requirement. Common JS is so 10 minutes ago.

**Reply by coreybutler** [reply]
@radcapitalist I'd summarize more like this:

- node-windows won't get much maintenance.
- It will eventually be shut down (won't take three years though).
- A replacement will be made.
- Unlike this project, the successor will have proper backing.

**Reply by radcapitalist** [reply]
Thanks for clarifying!

Eric
